crates/ruff_python_parser/src/string.rs

```diff
@@ -535,68 +227,95 @@ impl<'a> StringParser<'a> {
         }))
     }
 
-    fn parse(&mut self) -> Result<Vec<Expr>, LexicalError> {
-        if self.kind.is_any_fstring() {
-            self.parse_fstring(0)
-        } else if self.kind.is_any_bytes() {
-            self.parse_bytes().map(|expr| vec![expr])
+    fn parse(&mut self) -> Result<Expr, LexicalError> {
+        if self.kind.is_any_bytes() {
+            self.parse_bytes()
         } else {
-            self.parse_string().map(|expr| vec![expr])
+            self.parse_string()
         }
     }
 }
 
-fn parse_fstring_expr(source: &str, location: TextSize) -> Result<Expr, ParseError> {
-    let fstring_body = format!("({source})");
-    parse_expression_starts_at(&fstring_body, "<fstring>", location)
+pub(crate) fn parse_string_literal(
+    start: TextSize,
+    (source, kind, triple_quoted): (String, StringKind, bool),
+) -> Result<Expr, LexicalError> {
+    let start = start
+        + kind.prefix_len()
+        + if triple_quoted {
+            TextSize::from(3)
+        } else {
+            TextSize::from(1)
+        };
+    StringParser::new(source.as_str(), kind, start).parse()
 }
 
-fn parse_string(
-    source: &str,
-    kind: StringKind,
-    triple_quoted: bool,
+pub(crate) fn parse_fstring_middle(
     start: TextSize,
-) -> Result<Vec<Expr>, LexicalError> {
-    StringParser::new(source, kind, triple_quoted, start).parse()
+    (source, is_raw): (String, bool),
+) -> Result<Option<Expr>, LexicalError> {
+    // This is to account for the empty `FStringMiddle` token that is created
+    // to check for non-parenthesized lambda expressions.
+    if source.is_empty() {
+        return Ok(None);
+    }
+    let kind = if is_raw {
+        StringKind::RawString
+    } else {
+        StringKind::String
+    };
+    StringParser::new(source.as_str(), kind, start)
+        .parse_fstring_middle()
+        .map(Some)
 }
 
-pub(crate) fn parse_strings(
-    values: Vec<(TextSize, (String, StringKind, bool), TextSize)>,
-) -> Result<Expr, LexicalError> {
-    // Preserve the initial location and kind.
-    let initial_start = values[0].0;
-    let last_end = values.last().unwrap().2;
-    let initial_kind = (values[0].1 .1 == StringKind::Unicode).then(|| "u".to_owned());
-    let has_fstring = values
+/// Concatenate a list of string expressions into a single string expression.
+///
+/// This is mainly used for implicit string concatenation and the possible
+/// expression values are strings, bytes, and f-strings.
+pub(crate) fn concatenate_strings(
+    strings: Vec<ParenthesizedExpr>,
```

Comment:
What's the reason for passing `ParenthesizedExpr`? This should always either be StringLiteral, ByteLiteral, or FStringExpr nodes right? I recommend creating a new enum with just these three variants. It's very likely that that new union is smaller in size which helps performance.

Reply:
Yeah I actually thought of using that but it'll require moving the kind field (for unicode) into `StringConstant`. Giving a quick look at the usage of the kind field, it does not seem to be used many places so should be a quick refactor.

Reply:
PR: #7211